sklearn/model_selection/_search.py

```diff
@@ -513,84 +517,143 @@ def _fit(self, X, y, labels, parameter_iterable):
-            n_test_samples = 0
-            score = 0
-            all_scores = []
-            for this_score, this_n_test_samples, _, parameters in \
-                    out[grid_start:grid_start + n_splits]:
-                all_scores.append(this_score)
-                if self.iid:
-                    this_score *= this_n_test_samples
-                    n_test_samples += this_n_test_samples
-                score += this_score
-            if self.iid:
-                score /= float(n_test_samples)
-            else:
-                score /= float(n_splits)
-            scores.append((score, parameters))
-            # TODO: shall we also store the test_fold_sizes?
-            grid_scores.append(_CVScoreTuple(
-                parameters,
-                score,
-                np.array(all_scores)))
-        # Store the computed scores
-        self.grid_scores_ = grid_scores
+        all_test_scores, all_test_sample_counts, _ = zip(*out)
+
+        all_test_scores = np.array(
+            all_test_scores, dtype=np.float64).reshape(n_candidates, n_splits)
+        # NOTE test_sample counts (weights) remain the same for all candidates
+        test_sample_counts = np.array(all_test_sample_counts[:n_splits],
+                                      dtype=np.int)
+
+        scorer_name = _get_scorer_name(self.scorer_)
+        results = dict()
+
+        for split_i in range(n_splits):
+            results["%s_test_split_%d" % (scorer_name, split_i)] = (
+                all_test_scores[:, split_i])
+
+        # Computed the (weighted) mean and std for all the candidates
+        mean_w = np.average(
+            all_test_scores, axis=1,
+            weights=test_sample_counts if self.iid else None)
+
+        std_w = np.sqrt(np.average(
+            (all_test_scores - np.repeat(mean_w.reshape(n_candidates, 1),
+                                         n_splits, axis=1)) ** 2,
+            axis=1, weights=test_sample_counts if self.iid else None))
+
+        results["%s_test_mean" % scorer_name] = mean_w
+        results["%s_test_std" % scorer_name] = std_w
 
         # Find the best parameters by comparing on the mean validation score:
-        # note that `sorted` is deterministic in the way it breaks ties
-        best = sorted(grid_scores, key=lambda x: x.mean_validation_score,
-                      reverse=True)[0]
-        self.best_params_ = best.parameters
-        self.best_score_ = best.mean_validation_score
+        # note that `sorted` is preferred over ``np.argsort`` as it
+        # deterministically selects the candidate with the least param value
+        sorted_indices = sorted(
+            range(n_candidates),
+            key=results["%s_test_mean" % scorer_name].__getitem__,
+            reverse=True)
+        best_parameters = candidate_params[sorted_indices[0]]
+        results["%s_test_rank" % scorer_name] = np.empty((n_candidates,))
+        results["%s_test_rank" % scorer_name][sorted_indices] = (
```

Comment:
If going to report rank, we should be giving the same rank to tying models. [`scipy.stats.rankdata`](http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.stats.rankdata.html) will do this for you (and I think `method='min'` is fine here).